### PR TITLE
[tt-train] Fix memory usage underreporting with gradient accumulation

### PR DESCRIPTION
**Pull request by imichalakTT**
### PR Category

Bug fix

### Summary

When gradient accumulation was enabled, the memory tracker reported far less usage than the model actually consumed. This fix makes the reported numbers reflect true memory usage.

#### The bug

The tracker doesn't measure memory at a single point — it takes snapshots and stitches the per-segment deltas together into a cumulative total (each segment contributes what it allocated minus what it freed).

Gradient accumulation breaks the assumption that every micro-step has the same delta. On the first micro-step, the gradient buffers are allocated and then kept. So the first micro-step has a distinctly larger memory delta than the ones that follow.

The tracker saved every micro-step's forward/backward snapshot under the same name, so each overwrote the last and only the final micro-step survived. That surviving segment does not include the gradient-buffer allocation. When the tracker stitched its total from that one non-representative segment, it omitted the persistent gradient buffers entirely — undercounting exactly the memory that accumulation adds.

#### The fix

Snapshots are no longer overwritten. Every snapshot is now kept as its own segment, so the first micro-step's gradient-buffer allocation is preserved and correctly stitched into the total.

**Additional changes:**
- Unique per-micro-step labels (FORWARD_PASS_micro_0, FORWARD_PASS_micro_1, …). Not required for correctness now that nothing is overwritten.
- Lookups by snapshot name now fail loudly if a name is ambiguous, instead of quietly returning one arbitrary result. (No current caller is affected.)

#### Setup

A modified training_shakespeare_tinyllama.yaml config with batch size set to 2 and gradient accumulation steps to 4.

#### Previously

```
=== Memory Usage Summary ===

--- ENTRY ---
  DRAM: Segment Peak 0.00 MB, Allocations 0.00 MB, Deallocations 0.00 MB, Segment Change +0.00 MB
  DRAM: Cumulative Peak 0.00 MB, Cumulative Current 0.00 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- MODEL_CREATION ---
  DRAM: Segment Peak 2112.40 MB, Allocations 4203.03 MB, Deallocations 2098.50 MB, Segment Change +2104.53 MB
  DRAM: Cumulative Peak 2112.40 MB, Cumulative Current 2104.53 MB
  L1: Peak CB 0.29 MB, Peak Buffer 0.00 MB, Peak Total 0.29 MB
--- OPTIMIZER_CREATION ---
  DRAM: Segment Peak 4207.27 MB, Allocations 4207.27 MB, Deallocations 0.00 MB, Segment Change +4207.27 MB
  DRAM: Cumulative Peak 6311.80 MB, Cumulative Current 6311.80 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FORWARD_PASS ---
  DRAM: Segment Peak 7641.76 MB, Allocations 7767.05 MB, Deallocations 125.80 MB, Segment Change +7641.25 MB
  DRAM: Cumulative Peak 13953.56 MB, Cumulative Current 13953.05 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS ---
  DRAM: Segment Peak 516.00 MB, Allocations 15495.26 MB, Deallocations 23136.51 MB, Segment Change -7641.25 MB
  DRAM: Cumulative Peak 14469.05 MB, Cumulative Current 6311.80 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FORWARD_PASS ---
  DRAM: Segment Peak 7641.76 MB, Allocations 7767.05 MB, Deallocations 125.80 MB, Segment Change +7641.25 MB
  DRAM: Cumulative Peak 14469.05 MB, Cumulative Current 13953.05 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS ---
  DRAM: Segment Peak 516.00 MB, Allocations 15495.26 MB, Deallocations 23136.51 MB, Segment Change -7641.25 MB
  DRAM: Cumulative Peak 14469.05 MB, Cumulative Current 6311.80 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FORWARD_PASS ---
  DRAM: Segment Peak 7641.76 MB, Allocations 7767.05 MB, Deallocations 125.80 MB, Segment Change +7641.25 MB
  DRAM: Cumulative Peak 14469.05 MB, Cumulative Current 13953.05 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS ---
  DRAM: Segment Peak 516.00 MB, Allocations 15495.26 MB, Deallocations 23136.51 MB, Segment Change -7641.25 MB
  DRAM: Cumulative Peak 14469.05 MB, Cumulative Current 6311.80 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FORWARD_PASS ---
  DRAM: Segment Peak 7641.76 MB, Allocations 7767.05 MB, Deallocations 125.80 MB, Segment Change +7641.25 MB
  DRAM: Cumulative Peak 14469.05 MB, Cumulative Current 13953.05 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS ---
  DRAM: Segment Peak 516.00 MB, Allocations 15495.26 MB, Deallocations 23136.51 MB, Segment Change -7641.25 MB
  DRAM: Cumulative Peak 14469.05 MB, Cumulative Current 6311.80 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FIRST_ITERATION_COMPLETE ---
  DRAM: Segment Peak 0.18 MB, Allocations 0.18 MB, Deallocations 0.00 MB, Segment Change +0.18 MB
  DRAM: Cumulative Peak 14469.05 MB, Cumulative Current 6311.98 MB
  L1: Peak CB 0.06 MB, Peak Buffer 0.00 MB, Peak Total 0.06 MB

=== Final Totals ===
Overall DRAM Peak: 14469.05 MB, Final DRAM Usage: 6311.98 MB
```

#### Now

```
=== Memory Usage Summary ===

--- ENTRY ---
  DRAM: Segment Peak 0.00 MB, Allocations 0.00 MB, Deallocations 0.00 MB, Segment Change +0.00 MB
  DRAM: Cumulative Peak 0.00 MB, Cumulative Current 0.00 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- MODEL_CREATION ---
  DRAM: Segment Peak 2112.40 MB, Allocations 4203.03 MB, Deallocations 2098.50 MB, Segment Change +2104.53 MB
  DRAM: Cumulative Peak 2112.40 MB, Cumulative Current 2104.53 MB
  L1: Peak CB 0.29 MB, Peak Buffer 0.00 MB, Peak Total 0.29 MB
--- OPTIMIZER_CREATION ---
  DRAM: Segment Peak 4207.27 MB, Allocations 4207.27 MB, Deallocations 0.00 MB, Segment Change +4207.27 MB
  DRAM: Cumulative Peak 6311.80 MB, Cumulative Current 6311.80 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FORWARD_PASS_micro_0 ---
  DRAM: Segment Peak 7642.46 MB, Allocations 7767.51 MB, Deallocations 125.52 MB, Segment Change +7641.99 MB
  DRAM: Cumulative Peak 13954.26 MB, Cumulative Current 13953.79 MB
  L1: Peak CB 0.94 MB, Peak Buffer 0.00 MB, Peak Total 0.94 MB
--- BACKWARD_PASS_micro_0 ---
  DRAM: Segment Peak 500.09 MB, Allocations 13392.19 MB, Deallocations 18929.26 MB, Segment Change -5537.07 MB
  DRAM: Cumulative Peak 14453.88 MB, Cumulative Current 8416.72 MB
  L1: Peak CB 1.29 MB, Peak Buffer 0.00 MB, Peak Total 1.29 MB
--- FORWARD_PASS_micro_1 ---
  DRAM: Segment Peak 7641.76 MB, Allocations 7767.05 MB, Deallocations 125.80 MB, Segment Change +7641.25 MB
  DRAM: Cumulative Peak 16058.48 MB, Cumulative Current 16057.97 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS_micro_1 ---
  DRAM: Segment Peak 516.00 MB, Allocations 15495.26 MB, Deallocations 23136.51 MB, Segment Change -7641.25 MB
  DRAM: Cumulative Peak 16573.97 MB, Cumulative Current 8416.72 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FORWARD_PASS_micro_2 ---
  DRAM: Segment Peak 7641.76 MB, Allocations 7767.05 MB, Deallocations 125.80 MB, Segment Change +7641.25 MB
  DRAM: Cumulative Peak 16573.97 MB, Cumulative Current 16057.97 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS_micro_2 ---
  DRAM: Segment Peak 516.00 MB, Allocations 15495.26 MB, Deallocations 23136.51 MB, Segment Change -7641.25 MB
  DRAM: Cumulative Peak 16573.97 MB, Cumulative Current 8416.72 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FORWARD_PASS_micro_3 ---
  DRAM: Segment Peak 7641.76 MB, Allocations 7767.05 MB, Deallocations 125.80 MB, Segment Change +7641.25 MB
  DRAM: Cumulative Peak 16573.97 MB, Cumulative Current 16057.97 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- BACKWARD_PASS_micro_3 ---
  DRAM: Segment Peak 516.00 MB, Allocations 15495.26 MB, Deallocations 23136.51 MB, Segment Change -7641.25 MB
  DRAM: Cumulative Peak 16573.97 MB, Cumulative Current 8416.72 MB
  L1: Peak CB 0.00 MB, Peak Buffer 0.00 MB, Peak Total 0.00 MB
--- FIRST_ITERATION_COMPLETE ---
  DRAM: Segment Peak 0.18 MB, Allocations 0.18 MB, Deallocations 0.00 MB, Segment Change +0.18 MB
  DRAM: Cumulative Peak 16573.97 MB, Cumulative Current 8416.90 MB
  L1: Peak CB 0.06 MB, Peak Buffer 0.00 MB, Peak Total 0.06 MB

=== Final Totals ===
Overall DRAM Peak: 16573.97 MB, Final DRAM Usage: 8416.90 MB
```

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:imichalak/fix-memory-underreporting-with-grad-acc)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:imichalak/fix-memory-underreporting-with-grad-acc)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:imichalak/fix-memory-underreporting-with-grad-acc)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:imichalak/fix-memory-underreporting-with-grad-acc)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:imichalak/fix-memory-underreporting-with-grad-acc)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:imichalak/fix-memory-underreporting-with-grad-acc)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:imichalak/fix-memory-underreporting-with-grad-acc)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:imichalak/fix-memory-underreporting-with-grad-acc)

### CODE REVIEW

**Review by philei-tt — APPROVED** [review]
Thanks, good catch. Please add some tests

File: tt-train/sources/ttml/utils/memory_utils.cpp [hunk 1]
```diff
@@ -4,22 +4,53 @@
...
+using NamedTrace = std::pair<std::string, nlohmann::json>;
 
-static std::shared_ptr<ttnn::graph::GraphProcessor> graph_processor;
-static tt::tt_metal::IGraphProcessor::RunMode capture_mode = tt::tt_metal::IGraphProcessor::RunMode::NORMAL;
+// Global variables for memory usage tracking
+bool is_capture_active = false;
+std::shared_ptr<ttnn::graph::GraphProcessor> graph_processor;
+tt::tt_metal::IGraphProcessor::RunMode capture_mode = tt::tt_metal::IGraphProcessor::RunMode::NORMAL;
+
+std::vector<NamedTrace> traces;
```
[the remark this answers is not in the document]
Comment (imichalakTT) [reply tt-train/sources/ttml/utils/memory_utils.cpp:27]: added a regression test

**Review by abustamanteTT — APPROVED** [review]
